### DISCUSSION TITLE
[Enhancement]: Publishing message to Rabbitmq test-container Queue and consuming it

### DISCUSSION THREAD

**Discussion by miPlodder** [testcontainers/testcontainers-java#7453] [General]
### Module

RabbitMQ

### Proposal

- Currently, looks like we don't support publishing message to Rabbitmq test-container Queue and consuming it. 
- Also, when this functionality is completed, we should have proper documentation around it.

**Comment by eddumelendez** [comment]
> Currently, looks like we don't support publishing message to Rabbitmq test-container Queue and consuming it.

Can you elaborate on this, please? Are you talking about utility methods to use rabbimq command inside the container?

**Comment by miPlodder** [comment]
> Can you elaborate on this, please? Are you talking about utility methods to use rabbimq command inside the container?

I'm trying to write integration test for my Spring Application which uses Rabbitmq as Message Broker. In order to do that I have made below mentioned changes,

1. Adding Rabbitmq container details to the ConnectionFactory,

```
@Bean
  public CachingConnectionFactory connectionFactory() throws Exception {

    RabbitMQContainer rabbit =
        new RabbitMQContainer(DockerImageName.parse("rabbitmq:3.7.25-management-alpine"));
    rabbit.start();

    final RabbitConnectionFactoryBean rabbitConnectionFactoryBean =
        new RabbitConnectionFactoryBean();
    rabbitConnectionFactoryBean.setHost(rabbit.getHost());
    rabbitConnectionFactoryBean.setPort(rabbit.getAmqpPort());
    rabbitConnectionFactoryBean.setUsername(rabbit.getAdminUsername());
    rabbitConnectionFactoryBean.setPassword(rabbit.getAdminPassword());
    rabbitConnectionFactoryBean.setUseSSL(Boolean.parseBoolean(userSsl));
    rabbitConnectionFactoryBean.setSslAlgorithm(sslVersion);
    rabbitConnectionFactoryBean.setEnableHostnameVerification(false);
    rabbitConnectionFactoryBean.afterPropertiesSet();

    final CachingConnectionFactory connectionFactory =
        new CachingConnectionFactory(rabbitConnectionFactoryBean.getObject());

    return connectionFactory;
```

Similarly created RabbitTemplate, SimpleRabbitListenerContainerFactory  and using above mentioned connection factory

```
  @Bean
  public SimpleRabbitListenerContainerFactory rabbitListenerContainerFactory() throws Exception {
    SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
    factory.setConnectionFactory(connectionFactory());
    factory.setMaxConcurrentConsumers(1);
       return factory;
  }

 @Bean
  public RabbitTemplate amqpTemplate() throws Exception {
    return new RabbitTemplate(connectionFactory());
  }

@Bean
  public Queue abc() {
    return new Queue("abc", false);
  }

```

2. Next I'm creating Exchange and Routing Key binding to Queues,

```
 @Bean
  public Declarables httpBinding() {
    val queue = QueueBuilder.durable("queue").build();
    val declarables = new Declarables(exchange, queue,
        bind(queue).to("exchange")
            .with("routingKey"); 
    return declarables;
  }
```

3. Now, I send message to the Rabbitmq
```
rabbitTemplate.convertAndSend(exchange, routingKey, "data");
```

**But I don't see my Rabbitmq listener listening to this message here,**

```
 @RabbitListener(id = "1",
      queues = "queue")
  public void listener(final Message message) throws DimensionException {
  // .......
  }
```

- Am I missing something here, as per my understanding the only change that needs to be done is that `ConnectionFactory` should point to Rabbitmq test container, ryt?
- Also, can we do more operations with `RabbitMQContainer`?
- Even, I have bean created for the Queue, I'm getting this error in logs

```
2023-09-01 00:38:38,888 WARN  [1-1] listener.BlockingQueueConsumer (BlockingQueueConsumer.java:670) - Queue declaration failed; retries left=3
org.springframework.amqp.rabbit.listener.BlockingQueueConsumer$DeclarationException: Failed to declare queue(s):[abc]
```

**Reply by eddumelendez** [reply]
@miPlodder I've closed the issue due to doesn't fit on the issue backlog. Testcontainers `RabbitMQContainer` works well as you can see in [this example](https://github.com/eddumelendez/testcontainers-samples/blob/main/spring-boot-rabbitmq/src/test/java/com/example/consumer/SpringBootRabbitMQApplicationTests.java) and it Testcontainers' tests. Next time, please consider opening a discussion first or [joining the slack](https://slack.testcontainers.org/) to ask questions.

It's not clear for me if the container stopped running or what's happening as the information has not been shared. I'd recommend to take a look if the container is running or share a reproducible example. You can also take a look at RabbiMqAutoConfiguration in spring boot to create the factory correctly.

I think you need to create a `RabbitAdmin` bean in order to create the `Queue` defined in the bean, check the [spring-amqp docs](https://docs.spring.io/spring-amqp/reference/html/#with-xml-configuration)

> By default, the <rabbit:admin/> declaration automatically looks for beans of type Queue, Exchange, and Binding and declares them to the broker on behalf of the user. As a result, you need not use that bean explicitly in the simple Java driver. There are plenty of options to configure the properties of the components in the XML schema. You can use auto-complete features of your XML editor to explore them and look at their documentation.



> Also, can we do more operations with RabbitMQContainer?

No, we are not planning to do that. `rabbitmq` command can be performed using `container.execInContainer`. The reason is because it will couple the Testcontainers' RabbitMQ implementation to the image itself.

**Comment by miPlodder** [comment]
Not applicable
